### PR TITLE
sportstracker: init at 8.0.1

### LINKED ISSUE #298184: Package request: SportsTracker
**Project description**

SportsTracker is a desktop application for people who want to record and analyze their sporting activities. 

**Metadata**

* homepage URL: https://www.saring.de/sportstracker
* source URL: https://github.com/ssaring/sportstracker
* license: GPL-2.0 
* platforms: unix, linux, darwin, ...

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by TomaSajt**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/298184

This PR adds 1 package: `sportstracker`
I went with a capitalized executable name, because the releases also capitalize them.

This is a JavaFX application. The package wants to fetch `javafx` for itself, however we can provide `javafx` through nixpkgs by overriding the `jdk`. This way we won't have to patch any `.so` files inside the fetched files and there will be no platform-dependent `mvnHash` differences.

Unfortunately, currently there isn't a cached jdk21 with webkit enabled, so you'll have to wait out the long build time (~1 hour).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by TomaSajt** [comment]
@Thatoo could you test if this works for you?

**Comment by Thatoo** [comment]
How do I test? It doesn't appear on search.nixos.org ?

**Comment by TomaSajt** [comment]
> How do I test? It doesn't appear on search.nixos.org ?

It won't appear there unless this PR gets merged.

Until then you can either clone the repository and git checkout to my PR branch and run
```sh
nix-shell -I nixpkgs=. -p sportstracker
```
or
if you don't want to clone manually you can just run
```sh
nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/68e5e1732a328c9878a02cdece0a2b3403bfbec8.tar.gz -p sportstracker
```

Running this/these will temporarily make `SportsTracker` availiable as an executable in your current shell session.

**Comment by Thatoo** [comment]
Well, it launch but I can't say it works nicely.
When I try to open a file here is what I get 
image

In terminal here is what it says : 
```

[nix-shell:~]$ SportsTracker
2024-03-30 12:26:08 INFOS de.saring.sportstracker.gui.STDocumentImpl loadOptions Loading application options... 
2024-03-30 12:26:09 INFOS de.saring.sportstracker.gui.STControllerImpl$LoadTask call Loading application data... 
2024-03-30 12:26:09 INFOS de.saring.sportstracker.gui.STDocumentImpl readApplicationData Reading application data 
2024-03-30 12:26:09 INFOS de.saring.sportstracker.storage.db.DbStorage openDatabase Opening database jdbc:sqlite:/home/thatoo/.sportstracker/sportstracker.sqlite 
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
2024-03-30 12:26:09 INFOS de.saring.sportstracker.storage.db.DbStorage createSchema Creating database schema 
2024-03-30 12:26:09 INFOS de.saring.sportstracker.gui.STDocumentImpl readApplicationData Opened SQLite database in 513 msec 
2024-03-30 12:26:09 INFOS de.saring.sportstracker.storage.db.AbstractRepository readAll Reading all SportType entries 
2024-03-30 12:26:09 INFOS de.saring.sportstracker.storage.db.ExerciseRepository readAll Reading all Exercises 
2024-03-30 12:26:09 INFOS de.saring.sportstracker.storage.db.AbstractRepository readAll Reading all Note entries 
2024-03-30 12:26:09 INFOS de.saring.sportstracker.storage.db.AbstractRepository readAll Reading all Weight entries 
2024-03-30 12:26:09 INFOS de.saring.sportstracker.gui.STDocumentImpl readApplicationData Loaded all data in 6 msec 
2024-03-30 12:26:09 INFOS de.saring.sportstracker.gui.STDocumentImpl importApplicationDataFromXml Importing application data from XML 
2024-03-30 12:26:10 INFOS de.saring.sportstracker.storage.db.DbStorage importExistingApplicationData Importing existing application data to database 
2024-03-30 12:26:10 INFOS de.saring.sportstracker.storage.db.DbStorage commitChanges Committing database changes 
2024-03-30 12:26:10 INFOS de.saring.sportstracker.storage.db.AbstractRepository readAll Reading all SportType entries 
2024-03-30 12:26:10 INFOS de.saring.sportstracker.storage.db.ExerciseRepository readAll Reading all Exercises 
2024-03-30 12:26:10 INFOS de.saring.sportstracker.storage.db.AbstractRepository readAll Reading all Note entries 
2024-03-30 12:26:10 INFOS de.saring.sportstracker.storage.db.AbstractRepository readAll Reading all Weight entries 
2024-03-30 12:26:20 INFOS de.saring.sportstracker.gui.STControllerImpl onOpenHrmFile Opening HRM file '/home/Sport/Garmin/ACTIVITY/87KJ3909.FIT' in ExerciseViewer... 
2024-03-30 12:26:39 INFOS de.saring.sportstracker.gui.update.STUpdateChecker checkForUpdates Checking for SportsTracker updates... 
Killed
```

**Comment by TomaSajt** [comment]
> Well, it launch but I can't say it works nicely. When I try to open a file here is what I get

Could you tell me what system are you running on? (NixOS/MacOS/other)

**Comment by Thatoo** [comment]
```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.82, NixOS, 23.11 (Tapir), 23.11.5742.219951b495fc`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos-23.11, nixos-unstable"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```

**Comment by TomaSajt** [comment]
Okay, I pushed a patch for

**Comment by Thatoo** [comment]
It indeed open files now.
However, the tab "Chemin" (I guess "Path" or "Track" in english) don't display anything so we can't see the track on map : 
image

**Comment by TomaSajt** [comment]
Hm, I'll look i to it.
I don't have a `.fit` file to test with, what do you recommend I do?

**Comment by Thatoo** [comment]
Here is a .fit file (link available for one download only) : https://lufi.leprette.fr/r/wTpvHsfbwh#lmr8dkpNnlBjd5jvCME8DNy4TRAb29hY+Z40kNNI46A=

**Comment by TomaSajt** [comment]
It looks like `libjfxwebkit.so` isn't currently supported in the latest JDKs.
https://github.com/NixOS/nixpkgs/issues/162064

Not sure what I can do, but I'll try some things

**Comment by TomaSajt** [comment]
@Thatoo it looks like the program works correctly now, and displays the map as expected. I don't really recommend trying to build the package now because it takes about ~1 hour to compile everything.
I recommend waiting until the other PR reaches `nixos-unstable-small` (https://nixpk.gs/pr-tracker.html?pr=300573)

**Comment by Thatoo** [comment]
Is it ok now?

**Comment by TomaSajt** [comment]
> Is it ok now?

Should be, though I haven't built it in a while.

**Comment by Thatoo** [comment]
if it builds on unstable, can you also backport it to 24.05?

**Comment by TomaSajt** [comment]
This still isn't merged, if that's what you're asking.
I'll backport it to stable once it does get merged.

**Comment by TomaSajt** [comment]
Just checked, still works with rebased on latest unstable.

File: pkgs/by-name/sp/sportstracker/package.nix [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  makeDesktopItem,
+  copyDesktopItems,
+  wrapGAppsHook3,
+  gtk3,
+}:
+
+let
+  jdkWithJFX = jdk21.override {
+    enableJavaFX = true;
+    openjfx = openjfx21.override { withWebKit = true; };
```
Comment (FliegendeWurst) [inline pkgs/by-name/sp/sportstracker/package.nix:16]:
```suggestion
    openjfx21 = openjfx21.override { withWebKit = true; };
```

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

File: pkgs/by-name/sp/sportstracker/package.nix (hunk 1 above)
Comment (TomaSajt) [reply pkgs/by-name/sp/sportstracker/package.nix]: Thanks, fixed

**Review by FliegendeWurst — APPROVED** [review]

**Comment by Thatoo** [comment]
Could it be backported to 24.11?

File: pkgs/by-name/sp/sportstracker/package.nix [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  copyDesktopItems,
+  wrapGAppsHook3,
+  gtk3,
+}:
+
+let
+  jdkWithJFX = jdk21.override {
+    enableJavaFX = true;
+    openjfx21 = openjfx21.override { withWebKit = true; };
+  };
```
Comment (FliegendeWurst) [inline pkgs/by-name/sp/sportstracker/package.nix:17]: On Darwin this should probably just be `jdk21.override { enableJavaFX = true; }` (as is it breaks nixpkgs-review, but not GHA Eval).
My quick fix was to just throw if it isn't a real openjdk: https://github.com/NixOS/nixpkgs/pull/366232/files

**Pull request merged**